libbpf-tools: add cgroup_helpers and oomkill support display cgroup

Rtoax opened the pull request:
Using a simple test program (not shown) called OOM, we performed the
following two tests:

   1. Allocating unlimited memory
   2. Adding the process to a cgroup named oom-memcg and limiting its memory
      usage to 200MB.

When we do not print cgroup information, we can only see process information
and cannot see the difference between memcg and non-memcg.

    $ sudo ./oomkill
    Tracing OOM kills... Ctrl-C to stop.
    14:28:23 Triggered by PID 179201 ("oom"), OOM kill of PID 179201 ("oom"), 6114610 pages, loadavg: loadavg: 0.56 0.51 0.38 2/968 179204

    14:28:42 Triggered by PID 179212 ("oom"), OOM kill of PID 179212 ("oom"), 51200 pages, loadavg: loadavg: 0.40 0.47 0.37 3/968 179212

The function implemented by this patch can clearly display cgroup information.

    $ sudo ./oomkill  -c
    Tracing OOM kills... Ctrl-C to stop.
    14:32:59 Triggered by PID 179879 ("oom"), CGROUP 8309 ("/sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/session.slice/org.gnome.Shell@wayland.service"), OOM kill of PID 179879 ("oom"), 6114610 pages, loadavg: loadavg: 0.50 0.38 0.35 4/970 179879

    14:33:14 Triggered by PID 179884 ("oom"), CGROUP 122547 ("/sys/fs/cgroup/oom-memcg"), MEMCG 122547 ("/sys/fs/cgroup/oom-memcg"), OOM kill of PID 179884 ("oom"), 51200 pages, loadavg: loadavg: 0.47 0.38 0.35 3/971 179884
    
# and add `cgroup_helpers.c`.

yonghong-song commented on libbpf-tools/cgroup_helpers.c:
```diff
@@ -0,0 +1,291 @@
+			continue;
+
+		/* Only need cgroup or cgroup2 */
+		if (strcmp(fstype, "cgroup") && strcmp(fstype, "cgroup2"))
```
I don't have cgroup v1 environment. IIUC, cgroup v1 may have multiple mounts, but cgroup2 typically only have one (not considering containers). So that means the while loop in the above makes sense.

yonghong-song commented on libbpf-tools/cgroup_helpers.c:
```diff
@@ -0,0 +1,291 @@
+{
+	int i;
+
+	if (!roots || nentries <= 0)
```
Looks like 'nentries <= 0' is not needed. ir roots is not NULL, there should be at least one entry, right?

yonghong-song commented on libbpf-tools/cgroup_helpers.c:
```diff
@@ -0,0 +1,291 @@
+	return err ? -errno : st.st_ino;
+}
+
+long cgroup_cgroupid_of_mnt_path(const char *mntpoint, const char *cgroup_path)
```
Looks like this is not needed.

Rtoax replied:
Yes, thanks.

yonghong-song commented on libbpf-tools/cgroup_helpers.c:
```diff
@@ -0,0 +1,291 @@
+#include "cgroup_helpers.h"
+
+
+int cgroup_get_roots(char ***roots)
```
Looks like this should be a static function as it is only used in this file. The prototype in the cgroup_helpers.h is not needed. Please also check other functions.

yonghong-song commented on libbpf-tools/cgroup_helpers.c:
```diff
@@ -0,0 +1,291 @@
+	int dump_frequency;
+	int fsck_order;
+	FILE *fp;
+	int n = 0;
```
Please try to consolidate some variables with the same type.

yonghong-song commented on libbpf-tools/cgroup_helpers.c:
```diff
@@ -0,0 +1,291 @@
+	struct stat st;
+	size_t path_len;
+
+	if (!root || !match)
```
can 'match' be NULL?

Rtoax replied:
Replace with, thanks.

```diff
diff --git a/libbpf-tools/cgroup_helpers.c b/libbpf-tools/cgroup_helpers.c
index 894875426ae8..2f62737cc470 100644
--- a/libbpf-tools/cgroup_helpers.c
+++ b/libbpf-tools/cgroup_helpers.c
@@ -102,7 +102,9 @@ static int find_cgroup_from_root_recur(const char *root, match_fn match,
        struct stat st;
        size_t path_len;
 
-       if (!root || !match)
+       assert(match && "match_fn is NULL");
+
+       if (!root)
                return -EINVAL;
```

yonghong-song commented on libbpf-tools/cgroup_helpers.c:
```diff
@@ -0,0 +1,291 @@
+		return -ENOTDIR;
+
+	path = malloc(PATH_MAX);
+	assert(path && "Malloc failed");
```
Let us have proper return error code.

yonghong-song commented on libbpf-tools/cgroup_helpers.c:
```diff
@@ -0,0 +1,291 @@
+typedef int (*match_fn)(const char *path, void *arg);
+
+/**
+ * Recursively search for a matching cgroup in a root directory.
```
This is a little bit confusing. Here 'root' should be explicit saying cgroup mount root.

yonghong-song commented on libbpf-tools/cgroup_helpers.c:
```diff
@@ -0,0 +1,291 @@
+	return 0;
+}
+
+int cgroup_cgroup_path(long cgroupid, char *buf, size_t buf_len)
```
Maybe rename cgroup_cgroup_path to get_cgroupid_path?

yonghong-song commented on libbpf-tools/cgroup_helpers.c:
```diff
@@ -0,0 +1,291 @@
+	return found ? 0 : -ENOENT;
+}
+
+int cgroup_proc_for_each_cgroup_entry(pid_t pid, cgroup_proc_entry_fn callback,
```
This one seems not used.

yonghong-song commented on libbpf-tools/cgroup_helpers.h:
```diff
@@ -0,0 +1,74 @@
+#include <unistd.h>
+#include <sys/types.h>
+
+#define CGROUP_DEFAULT_MNTPOINT	"/sys/fs/cgroup/"
```
This one is not used.

yonghong-song commented on libbpf-tools/oomkill.h:
```diff
@@ -8,6 +8,8 @@ struct data_t {
 	__u32 tpid;
 	__u64 pages;
+	__u64 fcgroupid;
+	__u64 tcgroupid;
```
I would like to rename the above two to 'cgroupid' and 'mem_cgroupid'.

Rtoax commented:
> Could you actually show your simple test? This way, people can reproduce the issue easily.

Yes, thanks, the test code:

### oom_minimal.c

```c
#include <errno.h>
#include <malloc.h>
#include <unistd.h>
#include <stdio.h>

int main(void)
{
	void *mem;

	for (;;) {
		mem = malloc(getpagesize());
		if (!mem && errno == ENOMEM) {
			fprintf(stderr, "OOMing...\n");
		}
		*(int *)mem = 1;
		/* Just leak the memory */
	}
	return 0;
}
```

### cgroup-oom.sh

```bash
#!/bin/bash
set -e

readonly pid=$$
readonly CGROUP_NAME=oom-test

[[ -z ${OOMer} ]] && OOMer=oom_minimal

cleanup() {
	printf "\n"
	sudo rmdir /sys/fs/cgroup/${CGROUP_NAME}/
}
trap cleanup EXIT

sudo mkdir -p /sys/fs/cgroup/${CGROUP_NAME}/
echo ${pid} | sudo tee /sys/fs/cgroup/${CGROUP_NAME}/cgroup.procs
echo $((1024*1024*2)) | sudo tee /sys/fs/cgroup/${CGROUP_NAME}/memory.max
echo $((1024*1024*2)) | sudo tee /sys/fs/cgroup/${CGROUP_NAME}/memory.high

eval ./${OOMer} ${@}
```

### compile and run

```
gcc oom_minimal.c -o oom_minim
./cgroup-oom.sh
```

Rtoax commented:
@yonghong-song Thanks a lot, i just submit all changes and test code, please review, :)

Rtoax committed:
libbpf-tools: Add cgroup_helpers to get cgroup info

Some tools need to obtain cgroup information. For example, oomkill currently
only supports tracking process information and cannot obtain cgroup
information. It would be better if it could obtain memcg information.

For ease of maintenance, a separate commit is kept just for adding
cgroup_helpers.

Added interfaces:
    cgroup_cgroupid_of_path() - Get cgroupid from cgroup absolute path
    get_cgroupid_path() - Get cgroup path from cgroupid

Signed-off-by: Rong Tao <rongtao@cestc.cn>

Rtoax committed:
libbpf-tools: oomkill: support display memory cgroup

Using a simple test program (not shown) called OOM, we performed the
following two tests:

   1. Allocating unlimited memory
   2. Adding the process to a cgroup named oom-memcg and limiting its memory
      usage to 200MB.

When we do not print cgroup information, we can only see process information
and cannot see the difference between memcg and non-memcg.

    $ sudo ./oomkill
    Tracing OOM kills... Ctrl-C to stop.
    14:28:23 Triggered by PID 179201 ("oom"), OOM kill of PID 179201 ("oom"), 6114610 pages, loadavg: loadavg: 0.56 0.51 0.38 2/968 179204

    14:28:42 Triggered by PID 179212 ("oom"), OOM kill of PID 179212 ("oom"), 51200 pages, loadavg: loadavg: 0.40 0.47 0.37 3/968 179212

The function implemented by this patch can clearly display cgroup information.

    $ sudo ./oomkill  -c
    Tracing OOM kills... Ctrl-C to stop.
    14:32:59 Triggered by PID 179879 ("oom"), CGROUP 8309 ("/sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/session.slice/org.gnome.Shell@wayland.service"), OOM kill of PID 179879 ("oom"), 6114610 pages, loadavg: loadavg: 0.50 0.38 0.35 4/970 179879

    14:33:14 Triggered by PID 179884 ("oom"), CGROUP 122547 ("/sys/fs/cgroup/oom-memcg"), MEMCG 122547 ("/sys/fs/cgroup/oom-memcg"), OOM kill of PID 179884 ("oom"), 51200 pages, loadavg: loadavg: 0.47 0.38 0.35 3/971 179884

Link: https://github.com/iovisor/bcc/pull/5384
Signed-off-by: Rong Tao <rongtao@cestc.cn>

Rtoax commented:
I'm very sorry, because this PR is divided into two commits, one to add cgroup_helpers and the other to modify oomkill's cgroup support. My last push --force put the cgroup_helpers modification into the oomkill modification commit, so I have to resubmit it.

Rtoax commented:
Rebase to master.
